js/app/widgets/dynamicLogo.js

```diff
@@ -0,0 +1,339 @@
+
+const Knowledge = imports.app.knowledge;
+const WidgetSurfaceCache = imports.app.widgetSurfaceCache;
+
+const HEIGHT_WIDTH_RATIO = 1.5;
+
+const Width = {
+    MIN: 50,
+    MAX: 400,
+};
+
+const Height = {
+    MIN: 50,
+    MAX: 400,
+};
+
+/**
+ * Class: DynamicLogo
+ *
+ * This widget implements our dynamic logo system. A logo can be composed
+ * of an image, text or both. This widget is considered dynamic because it
+ * can be resized dynamically depending on its allocation size and it can
+ * switch between different layouts depending on the geometry of its
+ * allocation.
+ */
+const DynamicLogo = new Knowledge.Class({
+    Name: 'DynamicLogo',
+    Extends: Gtk.Widget,
+
+    Properties: {
+         /**
+         * Property: image-uri
+         *
+         * A URI to the image displayed in the logo.
+         */
+        'image-uri': GObject.ParamSpec.string('image-uri', 'image-uri', 'image-uri',
+            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT_ONLY, ''),
+         /**
+         * Property: text
+         *
+         * The text displayed in the logo.
+         */
+        'text': GObject.ParamSpec.string('text', 'text', 'text',
+            GObject.ParamFlags.READWRITE, ''),
+    },
+
+    StyleProperties: {
+        /**
+         * Property: mode
+         *
+         *
+         * -EknDynamicLogo-mode:
+         *  * 'full' - image and text are visible. Default.
+         *  * 'image' - only the image is visible.
+         *  * 'text' - only the text is visible.
+         */
+        'mode': GObject.ParamSpec.string('mode', 'mode', 'mode',
+            GObject.ParamFlags.READWRITE, 'full'),
+        /**
+         * Property: sizing
+         *
+         * -EknDynamicLogo-sizing:
+         *  * 'auto' - size determined by allocated size. Default.
+         *  * 'size-min' - size determined by min-width and min-height css properties.
+         */
+        'sizing': GObject.ParamSpec.string('sizing', 'sizing', 'sizing',
+            GObject.ParamFlags.READWRITE, 'auto'),
+        /**
+         * Property: text-transform
+         *
+         * -EknDynamicLogo-text-transform:
+         *  * 'none' - no transformation is performed. Default.
+         *  * 'uppercase' - the text is transformed to uppercase.
+         *  * 'lowercase' - the text is transformed to lowercase.
+         */
+        'text-transform': GObject.ParamSpec.string('text-transform', 'text-transform', 'text-transform',
+            GObject.ParamFlags.READWRITE, 'none'),
+    },
+
+    _init: function (props={}) {
+        this.parent(props);
+        this.set_has_window(false);
+
+        this._mode = 'full';
+        this._sizing = 'auto';
+        this._text_transform = 'none';
+
+        this._image = null;
+        let file = Gio.File.new_for_uri(this.image_uri);
+        if (file.query_exists(null)) {
+            let stream = file.read(null);
+            this._image = Rsvg.Handle.new_from_stream_sync(stream, file, 0, null);
+        }
```

Comment:
I did not realized it was construct only, so no in that case we do not need the setter.
BTW text property will also need to be construct only or have a setter since you need to queue a redraw when it changes.